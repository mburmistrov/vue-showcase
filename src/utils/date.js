import {
  differenceInDays as dfDiffInDays,
  isAfter as dfIsAfter,
  addDays as dfAddDays,
  subDays as dfSubDays,
  format as dfFormat
} from 'date-fns'

export function diffInDays(dateLeft, dateRight) {
  return dfDiffInDays(dateLeft, dateRight)
}

export function isDateAfter(date, dateToCompare) {
  return dfIsAfter(date, dateToCompare)
}

export function addDays(date, amount) {
  return dfAddDays(date, amount)
}

export function subDays(date, amount) {
  return dfSubDays(date, amount)
}

export function format(date, format = 'dd.MM.yyyy') {
  return dfFormat(date, format)
}
