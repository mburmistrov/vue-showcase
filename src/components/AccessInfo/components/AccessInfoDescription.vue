<script>
import { mapState, mapGetters } from 'vuex'
import { format } from '@/utils/date'

export default {
  computed: {
    ...mapState(['currentUser']),
    ...mapGetters(['isDemo', 'licenseValidUntil']),
    text() {
      return this.isDemo
        ? this.$t('accessInfo.desc.demoText')
        : this.$t('accessInfo.desc.text', { licenseUntil: format(this.licenseValidUntil) })
    },
    actionText() {
      return this.isDemo ? this.$t('accessInfo.desc.demoActionText') : this.$t('accessInfo.desc.actionText')
    }
  },
  methods: {
    goToPayment() {
      this.$emit('goToPayment')
    }
  }
}
</script>
<template lang="pug">
.access-info-description
  .access-info-description__text(v-html="text")
  .access-info-description__action(ref="goToPayment" @click="goToPayment") {{ actionText }}
</template>
<style lang="sass">
@import "~/src/styles/variables"
@import "~/src/styles/elevation"
.access-info-description
  position: absolute
  color: black
  padding: 10px
  border-radius: 3px
  +elevation
  .access-info-description__text
    font-size: 12px
    line-height: 1.6
  .access-info-description__action
    background-color: $c-primary
    color: white
    box-sizing: border-box
    border-radius: 3px
    padding: 5px
    margin-top: 10px
    text-align: center
    cursor: pointer
</style>
