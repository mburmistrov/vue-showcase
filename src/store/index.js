import Vuex from 'vuex'
import { isDateAfter } from '@/utils/date'

let store = null

export const storeConfig = Object.freeze({
  state: {
    currentUser: null
  },
  getters: {
    licenseValidUntil: (state) => {
      if (!state.currentUser) return null
      return state.currentUser.licenseValidUntil
    },
    hasValidLicense: (state, getters) => {
      if (!getters.licenseValidUntil) return false
      return isDateAfter(getters.licenseValidUntil, new Date())
    },
    isDemo: (state, getters) => {
      return !getters.hasValidLicense
    }
  },
  mutations: {
    setCurrentUser(state, currentUser) {
      state.currentUser = currentUser
    }
  }
})

export function createStore() {
  store = store || new Vuex.Store(storeConfig)
  return store
}
