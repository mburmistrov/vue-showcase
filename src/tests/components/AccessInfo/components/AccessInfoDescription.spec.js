import { shallowMount, createLocalVue } from '@vue/test-utils'
import AccessInfoDescription from '@/components/AccessInfo/components/AccessInfoDescription'
import { merge } from 'lodash'
import Vuex from 'vuex'
import { storeConfig } from '@/store'
import { getUser, withValidLicense, withExpiredLicense, withoutLicense } from '@/tests/utils/getUser'

const localVue = createLocalVue()
localVue.use(Vuex)

function createStore(overrides) {
  const defaultStoreConfig = {
    state: {
      currentUser: getUser([withValidLicense])
    },
    getters: storeConfig.getters
  }
  return new Vuex.Store(
    merge(defaultStoreConfig, overrides)
  )
}

describe('AccessInfoDescription', () => {
  function createWrapper(overrides) {
    const defaultMountingOptions = {
      localVue,
      store: createStore()
    }

    return shallowMount(AccessInfoDescription, merge(defaultMountingOptions, overrides))
  }

  function createWrapperWithCurrentUser(user) {
    const store = createStore({
      state: {
        currentUser: user
      }
    })
    return createWrapper({ store })
  }

  describe('template', () => {
    it('should emit goToPayment on goToPayment click', () => {
      const wrapper = createWrapper()
      wrapper.findComponent({ ref: 'goToPayment' }).trigger('click')
      expect(wrapper.emitted().goToPayment).toBeTruthy()
    })

    describe('snapshots', () => {
      it('should match snapshot if user has no license', () => {
        const wrapper = createWrapperWithCurrentUser(getUser([withoutLicense]))
        expect(wrapper.element).toMatchSnapshot()
      })

      it('should match snapshot if user has expired license', () => {
        const wrapper = createWrapperWithCurrentUser(getUser([withExpiredLicense]))
        expect(wrapper.element).toMatchSnapshot()
      })

      it('should match snapshot if user has valid license', () => {
        const wrapper = createWrapperWithCurrentUser(getUser([withValidLicense]))
        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })
})
