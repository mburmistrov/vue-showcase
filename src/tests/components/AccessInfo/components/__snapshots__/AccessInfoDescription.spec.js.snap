// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessInfoDescription template snapshots should match snapshot if user has expired license 1`] = `
<div
  class="access-info-description"
>
  <div
    class="access-info-description__text"
  >
    accessInfo.desc.demoText
  </div>
  <div
    class="access-info-description__action"
  >
    accessInfo.desc.demoActionText
  </div>
</div>
`;

exports[`AccessInfoDescription template snapshots should match snapshot if user has no license 1`] = `
<div
  class="access-info-description"
>
  <div
    class="access-info-description__text"
  >
    accessInfo.desc.demoText
  </div>
  <div
    class="access-info-description__action"
  >
    accessInfo.desc.demoActionText
  </div>
</div>
`;

exports[`AccessInfoDescription template snapshots should match snapshot if user has valid license 1`] = `
<div
  class="access-info-description"
>
  <div
    class="access-info-description__text"
  >
    accessInfo.desc.text
  </div>
  <div
    class="access-info-description__action"
  >
    accessInfo.desc.actionText
  </div>
</div>
`;
